gdshowreelvote/auth.py

```diff
@@ -0,0 +1,157 @@
+STAFF_ROLE = 'staff'
+
+oauth = OAuth()
+
+MOCK_USERS = {
+    'moderator': {'mod': True, 'staff': True, 'fund_member': True},
+    'staff': {'mod': False, 'staff': True, 'fund_member': False},
+    'diamond-member': {'mod': False, 'staff': False, 'fund_member': True},
+    'user': {'mod': False, 'staff': False, 'fund_member': False},
+}
+
+
+def login_required(f):
+    @wraps(f)
+    def decorated_func(*args, **kwargs):
+        if session.get('user'):
+            return f(*args, **kwargs)
+        else:
+            return redirect(url_for('oidc.login'))
+    return decorated_func
+
+
+def admin_required(f):
+    @wraps(f)
+    def decorated_func(*args, **kwargs):
+        if session.get('user') and ADMIN_ROLE in session['user'].get('roles', []):
+            return f(*args, **kwargs)
+        else:
+            return redirect(url_for('oidc.login'))
+    return decorated_func
+
+
+def _can_vote(user: Dict) -> bool:
+    if ADMIN_ROLE in user.get('roles', []) or STAFF_ROLE in user.get('roles', []) or _fund_member_can_vote(user):
+        return True
+    return False
+    
+
+def vote_role_required(f):
+    @wraps(f)
+    def decorated_func(*args, **kwargs):
+        if session.get('user') and _can_vote(session['user']):
+            return f(*args, **kwargs)
+        else:
+            return redirect(url_for('oidc.login'))
+    return decorated_func
+
+
+def get_issuer():
+    if current_app.config.get('OIDC_MOCK', False):
+        return 'https://example.org/keycloak/realms/test'
+    else:
+        return oauth.oidc.load_server_metadata()['issuer']
+
+
+# Mock implementation
+
+def mock_login():
+    content = render_template('mock-login.html', users=MOCK_USERS)
+    return render_template('default.html', content=content, title='Login')
+
+
+def mock_auth():
+    username = request.form.get('username', '').lower()
+    if not username in MOCK_USERS:
+        return redirect(url_for('oidc.login'))
+    roles = []
+    roles.append(ADMIN_ROLE) if MOCK_USERS[username]['mod'] else None
+    roles.append(STAFF_ROLE) if MOCK_USERS[username]['staff'] else None
+    fund_roles = []
+    fund_roles.append('tier-diamond') if MOCK_USERS[username]['fund_member'] else None
+    oidc_info = {
+        'sub': f'MOCK_USER:{username}',
+        'email_verified': True,
+        'name': username.capitalize(),
+        'preferred_username': username,
+        'given_name': username.capitalize(),
+        'family_name': username.capitalize(),
+        'email': f'{username}@example.com',
+        'roles': roles,
+        'fund': {'roles': fund_roles}
+    }
+    user = DB.session.get(User, oidc_info['sub'])
+    if not user:
+        user = User(id=oidc_info['sub'], username=oidc_info['name'], email=oidc_info['email'], 
+                    is_staff=STAFF_ROLE in oidc_info['roles'] or _fund_member_can_vote(oidc_info), 
+                    is_superuser=ADMIN_ROLE in oidc_info['roles'])
+        DB.session.add(user)
+        DB.session.commit()
```

Comment:
I do wonder if we can move the user creation into its own method and instead use correct OIDC info here. Then the `mock_auth()` and `oidc_auth()` methods can both just call that function. Something like `update_or_create_user(oidc_info)`